mempool/src/shared_mempool/network.rs

```diff
@@ -131,17 +152,372 @@ pub(crate) struct MempoolNetworkInterface {
+            return true;
+        }
+
+        // Outbound connections are upstream on non-P2P networks
+        if let Some(metadata) = metadata {
+            metadata.origin == ConnectionOrigin::Outbound
+        } else {
+            self.sync_states.read(peer).is_some()
+        }
+    }
+
+    pub fn process_broadcast_ack(
+        &self,
+        peer: PeerNetworkId,
+        request_id_bytes: Vec<u8>,
+        retry: bool,
+        backoff: bool,
+        timestamp: SystemTime,
+    ) {
+        let batch_id = if let Ok(id) = bcs::from_bytes::<BatchId>(&request_id_bytes) {
+            id
+        } else {
+            counters::invalid_ack_inc(&peer, counters::INVALID_REQUEST_ID);
+            return;
+        };
+
+        let mut sync_states = self.sync_states.write_lock();
+
+        let sync_state = if let Some(state) = sync_states.get_mut(&peer) {
+            state
+        } else {
+            counters::invalid_ack_inc(&peer, counters::UNKNOWN_PEER);
+            return;
+        };
+
+        if let Some(sent_timestamp) = sync_state.broadcast_info.sent_batches.remove(&batch_id) {
+            let rtt = timestamp
+                .duration_since(sent_timestamp)
+                .expect("failed to calculate mempool broadcast RTT");
+
+            let network_id = peer.network_id();
+            let peer_id = peer.peer_id().short_str();
+            counters::SHARED_MEMPOOL_BROADCAST_RTT
+                .with_label_values(&[network_id.as_str(), peer_id.as_str()])
+                .observe(rtt.as_secs_f64());
+
+            counters::shared_mempool_pending_broadcasts(&peer).dec();
+        } else {
+            trace!(
+                LogSchema::new(LogEntry::ReceiveACK)
+                    .peer(&peer)
+                    .batch_id(&batch_id),
+                "batch ID does not exist or expired"
+            );
+            return;
+        }
+
+        trace!(
+            LogSchema::new(LogEntry::ReceiveACK)
+                .peer(&peer)
+                .batch_id(&batch_id)
+                .backpressure(backoff),
+            retry = retry,
+        );
+        tasks::update_ack_counter(&peer, counters::RECEIVED_LABEL, retry, backoff);
+
+        if retry {
+            sync_state.broadcast_info.retry_batches.insert(batch_id);
+        }
+
+        // Backoff mode can only be turned off by executing a broadcast that was scheduled
+        // as a backoff broadcast.
+        // This ensures backpressure request from remote peer is honored at least once.
+        if backoff {
+            sync_state.broadcast_info.backoff_mode = true;
+        }
+    }
+    pub fn is_backoff_mode(&self, peer: &PeerNetworkId) -> bool {
+        if let Some(state) = self.sync_states.write_lock().get(peer) {
+            state.broadcast_info.backoff_mode
+        } else {
+            // If we don't have sync state, we shouldn't backoff
+            false
+        }
+    }
+
+    pub fn execute_broadcast<V>(
```

Comment:
Not a comment for you, but, wow: this method is long..

Reply:
Yeah, I had to restrain myself as to not change functionality along with this refactor.  I want to come back and fix some of these ridiculous pieces soon